lib/gateway.cc

```diff
@@ -0,0 +1,159 @@
+#include <sys/param.h>
+#include <unistd.h>
+
+#include "prometheus/client_metric.h"
+#include "prometheus/gateway.h"
+#include "prometheus/serializer.h"
+#include "prometheus/text_serializer.h"
+
+namespace prometheus {
+
+const char* CONTENT_TYPE = "text/plain; version=0.0.4; charset=utf-8";
+
+Gateway::Gateway(const std::string& uri, const std::string jobname,
+                 const labels_t* labels, const std::string username,
+                 const std::string password)
+    : uri_(uri), jobname_(jobname), username_(username), password_(password) {
+  if (labels) {
+    std::stringstream ss;
+
+    if (labels) {
+      for (auto& label : *labels) {
+        ss << "/" << label.first << "/" << label.second;
+      }
+    }
+
+    labels_ = ss.str();
+  }
+}
+
+const Gateway::labels_t Gateway::instance_label(void) {
+  char hostname[MAXHOSTNAMELEN] = {0};
+
+  if (gethostname(hostname, MAXHOSTNAMELEN - 1)) {
+    hostname[0] = 0;
+  }
+
+  return Gateway::labels_t{{"instance", hostname}};
+}
+
+void Gateway::RegisterCollectable(const std::weak_ptr<Collectable>& collectable,
+                                  const labels_t* labels) {
+  std::stringstream ss;
+
+  if (labels) {
+    for (auto& label : *labels) {
+      ss << "/" << label.first << "/" << label.second;
+    }
+  }
+
+  collectables_.push_back(std::make_pair(collectable, ss.str()));
+}
+
+std::stringstream Gateway::job_uri(void) const {
```

Comment:
why not return a string on this method?

Reply:
I now calculate the value in the ctor. See #124